### PR TITLE
Added a MAX7219 device

### PR DESCRIPTION
**Pull request by schonbachler**

### CODE REVIEW

File: src/devices/Max7219/samples/Max7219.sample.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            {
+                ClockFrequency = 10_000_000,
+                Mode = SpiMode.Mode0
+            };
+            var spi = new UnixSpiDevice(connectionSettings);
+            var devices = new Max7219(spi, cascadedDevices: 4);
+            using (devices)
+            {
+                Console.WriteLine("Display-Test");
+                devices.SetRegister(Max7219.Register.DISPLAYTEST, 1);
```
Comment (joperezr) [inline src/devices/Max7219/samples/Max7219.sample.cs]: Why is this needed? If it is actually needed for printing messages, should we have this as part of the Init() method instead?

File: src/devices/Max7219/samples/Max7219.sample.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                Thread.Sleep(1000);
+                Console.WriteLine("Init");
+                devices.Init();
+                foreach (Max7219.RotationType rotation in Enum.GetValues(typeof(Max7219.RotationType)))
```
Comment (joperezr) [inline src/devices/Max7219/samples/Max7219.sample.cs]: Can we add comments in the sample code in order to understand what are the different values of RotationType?

File: src/devices/Max7219/samples/Max7219.sample.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                {
+                    Thread.Sleep(500);
+                    devices.Rotation = rotation;
+                    Console.WriteLine($"Rotation {devices.Rotation}");
+                    for (var loop = 0; loop < 2; loop++)
```
Comment (joperezr) [inline src/devices/Max7219/samples/Max7219.sample.cs]: Same here, we should add some comments in the code on what is this 4 nested for-loop doing so that it is more  obvious for consumers on how to use the binding.

File: src/devices/Max7219/samples/Max7219.sample.csproj [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="System.Device.Gpio" Version="0.1.0-prerelease*" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="$(MSBuildThisFileDirectory)\..\Max7219.csproj" />
```
Comment (joperezr) [inline src/devices/Max7219/samples/Max7219.sample.csproj]: NIT: Can we change the backslashes to forward slashes instead? This is sometimes a problem when building in Linux because backslash is a valid character for a folder name.

File: src/devices/Max7219/samples/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
+# Write to a 8x8 Dot Matrix Module
+
+You can use .NET Core to drive MAX7219 based DOt Matrix Modules.
```
Comment (joperezr) [inline src/devices/Max7219/samples/README.md]: NIT: `Dot` instead of `DOt`

File: src/devices/Max7219/samples/README.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+## How to Cross Compile and Run this sample
+
+This example can also be cross-compiled on another machine and then executed on the Raspberry PI. This can be achieved with enabled SSH access to the RaspPi as follows.
+
+* Publish project on the development machine
+
+        cd ~/Projects/iot/src/devices/Max7219/samples
+        dotnet publish -c Release -r linux-arm
+
+* Synchronize published folder to the RaspPi via rsync over ssh
```
Comment (joperezr) [inline src/devices/Max7219/samples/README.md]: NIT: perhaps add a comment that you can use `pscp` instead when developing on a Windows machine

File: src/devices/Max7219/samples/README.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+        rsync -avz -e 'ssh' bin/Release/netcoreapp2.1/linux-arm/publish/  pi@192.168.1.192:/home/pi/max-sample/
+
+* Execute the program on the RaspPi or remote via SSH
+
+        ssh pi@192.168.1.192
+        cd /home/pi/max-sample/
+        ./max-sample
```
Comment (joperezr) [inline src/devices/Max7219/samples/README.md]: In order for this to work, the app must be a stand-alone app which your sample is not at the moment. If you want that to work, can you add the following properties to the sample:
```xml
<PropertyGroup>
  <RuntimeIdentifier>linux-arm</RuntimeIdentifier>
  <SelfContained>true</SelfContained>
</PropertyGroup>
```

File: src/devices/Max7219/Max7219.cs [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+namespace Iot.Device
+{
+    public class Max7219 : IDisposable
+    {
+        private SpiDevice _spiDevice;
+
+        /// <summary>
+        /// Number of digits Register per Module
+        /// </summary>
+        const int NumDigits = 8;
```
Comment (joperezr) [inline src/devices/Max7219/Max7219.cs]: Should this be private?

File: src/devices/Max7219/Max7219.cs [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+
+        /// <summary>
+        /// Internal buffer used to write to the devices.
+        /// </summary>
+        private readonly byte[] _writeBuffer;
+
+        /// <summary>
+        /// A Buffer that contains the values of the digits registers per device
+        /// </summary>
+        private readonly byte[,] _buffer;
```
Comment (joperezr) [inline src/devices/Max7219/Max7219.cs]: It is not obvious to me what is the difference between the write buffer and the _buffer, can we consider renaming this one and adding more helpful comment?

File: src/devices/Max7219/Max7219.cs [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+        public void Brightness(int intensity)
+        {
+            if (intensity < 0 || intensity > 15)
+                throw new ArgumentOutOfRangeException(nameof(intensity), $"Invalid intensity for Brightness {intensity}");
+            SetRegister(Register.INTENSITY, (byte)intensity);
+        }
+
+        /// <summary>
+        /// Gets or Sets the value to the digit value for a given device
+        /// and digit position
```
Comment (joperezr) [inline src/devices/Max7219/Max7219.cs]: What does digit mean on this context?

File: src/devices/Max7219/Max7219.cs [hunk 11]
```diff
@@ -0,0 +1,323 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+using System;
+using System.Device.Gpio;
```
Comment (joperezr) [inline src/devices/Max7219/Max7219.cs]: Correct me if I'm wrong but I don't believe you use GPIO at all here right? It looks like everything is done through SPI.

File: src/devices/Max7219/Max7219.cs [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+        public void Dispose()
+        {
+            if (_spiDevice != null)
+            {
+                _spiDevice.Dispose();
+                _spiDevice = null;
+            }
+        }
+
+        public enum Register : byte
```
Comment (joperezr) [inline src/devices/Max7219/Max7219.cs]: What we have been doing for now on other device bindings is to have the namespace be something like: Iot.Device.Max7219 and that way you can have these two enums not be nested types but instead be directly on the binding itself. That way you can also move them into different files which will make this look cleaner.

**Comment by joperezr** [comment]
Left some comments but this is looking good overall. Thanks a lot for your contribution!

File: src/devices/Max7219/samples/README.md (hunk 7 above)
Comment (schonbachler) [reply src/devices/Max7219/samples/README.md]: This is not true. It works as described on my machine (macos 10.13.6). 
I think 
```xml
         <OutputType>Exe</OutputType>
```
in combination with -r linux-arm option does the trick.

**Comment by schonbachler** [comment]
I will consider the suggestions and deliver an update in a few days.

File: src/devices/Max7219/samples/README.md (hunk 7 above)
Comment (joperezr) [reply src/devices/Max7219/samples/README.md]: I believe that what happens is that wherever you are running it (on a Raspberry Pi presumably) you have already installed .NET Core. The advantage of my suggestion of setting those two properties, is that you won't require the Raspberry Pi to have .NET Core already installed.

**Comment by joperezr** [comment]
> I will consider the suggestions and deliver an update in a few days.

Sounds good to me! 👍

File: src/devices/Max7219/Font.cs [hunk 13]
```diff
@@ -0,0 +1,41 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+namespace Iot.Device.Max7219
+{
+    public class Font
```
Comment (joperezr) [inline src/devices/Max7219/Font.cs]: It's not clear to me why we have a Font class and a Fonts class as well. Do we really need both? If so, can we have better naming explaining what they represent and perhaps also adding summary comments above each class?

File: src/devices/Max7219/Font.cs [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        readonly byte[][] _bitmap;
```
Comment (joperezr) [inline src/devices/Max7219/Font.cs]: NIT: We usually add the access modifier explicitly for private members.

**Comment by joperezr** [comment]
>         public static readonly Font CP437 = new Font(

Can we get a link of where to find the definitions for these fonts? also perhaps include how to add additional fonts?

---
Refers to: src/devices/Max7219/Fonts.cs:11 in 1b7d3da. [](commit_id = 1b7d3da640ac475d435eca9f0a0ed3d876698a4e, deletion_comment = False)

**Comment by joperezr** [comment]
>         public static readonly Font SINCLAIR_FONT = new Font(

We should change Casing to use CamelCase instead to match classes naming.

---
Refers to: src/devices/Max7219/Fonts.cs:278 in 1b7d3da. [](commit_id = 1b7d3da640ac475d435eca9f0a0ed3d876698a4e, deletion_comment = False)

**Comment by joperezr** [comment]
>     }

NIT: Extra spaces here.

---
Refers to: src/devices/Max7219/Fonts.cs:1337 in 1b7d3da. [](commit_id = 1b7d3da640ac475d435eca9f0a0ed3d876698a4e, deletion_comment = False)

File: src/devices/Max7219/MatrixTextWriter.cs [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Linq;
+using System.Threading;
+
+namespace Iot.Device.Max7219
+{
+
+    /// <summary>
```
Comment (joperezr) [inline src/devices/Max7219/MatrixTextWriter.cs]: NIT: Extra space

File: src/devices/Max7219/MatrixTextWriter.cs [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+    /// Matrix text writer.
+    /// </summary>
+    public class MatrixTextWriter
```
Comment (joperezr) [inline src/devices/Max7219/MatrixTextWriter.cs]: Is this the main class that consumers will interact with? or do we expect people to use Max7219 class directly?

File: src/devices/Max7219/Font.cs [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+        byte[] Trim(byte[] bitmap)
+        {
+            var start = 0;
+            var end = bitmap.Length;
+            while (start < end && bitmap[start] == 0)
+                start++;
+            while (end > start && bitmap[end - 1] == 0)
+                end--;
+            if (start > 0 || end < bitmap.Length)
+                return new Span<byte>(bitmap).Slice(start, end - start).ToArray();
```
Comment (tarekgh) [inline src/devices/Max7219/Font.cs]: why we don't cache this and always allocate? this will be very costy

File: src/devices/Max7219/Font.cs [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+        readonly byte[][] _bitmap;
+
+        public Font(params byte[][] bitmap)
+        {
+            _bitmap = bitmap;
+        }
+
+        public byte[] GetItem(char chr)
+        {
+            var bitmap = _bitmap[chr];
```
Comment (tarekgh) [inline src/devices/Max7219/Font.cs]: >var bitmap = _bitmap[chr]; [](start = 12, length = 26)

is it ok to get out of range exception here? I feel the design of this class is not good enough.

File: src/devices/Max7219/Font.cs [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+using System;
+
+namespace Iot.Device.Max7219
+{
+    public class Font
+    {
+        readonly byte[][] _bitmap;
+
+        public Font(params byte[][] bitmap)
```
Comment (tarekgh) [inline src/devices/Max7219/Font.cs]: >params [](start = 20, length = 6)

why you have params keyword here?

**Comment by tarekgh** [comment]
>         public static readonly Font CP437 = new Font(

could you please write a comment telling how the font is going to be used and what are the values inside the font arrays? are these bitmap values of the characters or just mapping value to system font?

---
Refers to: src/devices/Max7219/Fonts.cs:11 in 1b7d3da. [](commit_id = 1b7d3da640ac475d435eca9f0a0ed3d876698a4e, deletion_comment = False)

**Comment by tarekgh** [comment]
>         new byte[] { 0x7E, 0x81, 0x95, 0xB1, 0xB1, 0x95, 0x81, 0x7E },  // 0x01

this will be very slow initializing. did you consider just have one dimension array and do the the calculation when accessing it? if we make it one dimension array we'll allocate just one object.

---
Refers to: src/devices/Max7219/Fonts.cs:13 in 1b7d3da. [](commit_id = 1b7d3da640ac475d435eca9f0a0ed3d876698a4e, deletion_comment = False)

File: src/devices/Max7219/MatrixTextWriter.cs [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+    {
+        readonly Max7219 _device;
+
+        public MatrixTextWriter(Max7219 device, Font font)
+        {
+            _device = device;
```
Comment (tarekgh) [inline src/devices/Max7219/MatrixTextWriter.cs]: >_device [](start = 12, length = 7)

any validation needed for the device here? like passing null for instance?

File: src/devices/Max7219/MatrixTextWriter.cs [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+        /// <summary>
+        /// Writes a char to the given device with the specified font.
+        /// </summary>
+        public void WriteLetter(int deviceId, char chr, bool flush = true)
+        {
+            var bitmap = Font.GetItem(chr);
+            var end = Math.Min(bitmap.Length, Max7219.NumDigits);
+            for (int col = 0; col < end; col++)
+            {
+                _device[deviceId, col] = bitmap[col];
```
Comment (tarekgh) [inline src/devices/Max7219/MatrixTextWriter.cs]: >_device [](start = 16, length = 7)

could you please add a comment regarding the _device buffer, how the buffer mapped to the device display

File: src/devices/Max7219/MatrixTextWriter.cs [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+            if (flush)
+            {
+                _device.Flush();
+            }
+        }
+
+        /// <summary>
+        ///  Scrolls the underlying buffer (for all cascaded devices) up one pixel
+        /// </summary>
+        public void ScrollUp(bool flush)
```
Comment (tarekgh) [inline src/devices/Max7219/MatrixTextWriter.cs]: >flush [](start = 34, length = 5)

why flush parameter not defaulted to true as we did in WriteLetter?

Comment (tarekgh) [reply src/devices/Max7219/MatrixTextWriter.cs]: this comment apply to the rest of the methods takes flush

---
In reply to: [245435432](https://github.com/dotnet/iot/pull/104#discussion_r245435432) [](ancestors = 245435432)

File: src/devices/Max7219/MatrixTextWriter.cs [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+
+        /// <summary>
+        /// Shows a message on the device. 
+        /// If it's longer then the total width (or <see paramref="alwaysScroll"/> == true), 
+        /// it transitions the text message across the devices from right-to-left.
+        /// </summary>
+        public void ShowMessage(string text, float delay = 0.05f, bool alwaysScroll = false)
+        {
+            var displayLength = Max7219.NumDigits * _device.CascadedDevices;
+            var src = text.Select(Font.GetItem);
```
Comment (tarekgh) [inline src/devices/Max7219/MatrixTextWriter.cs]: >var [](start = 12, length = 3)

please use the type name instead of var as it is not obvious what type would be

File: src/devices/Max7219/MatrixTextWriter.cs [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+                _device.Flush();
+            }
+        }
+
+        /// <summary>
+        /// Shows a message on the device. 
+        /// If it's longer then the total width (or <see paramref="alwaysScroll"/> == true), 
+        /// it transitions the text message across the devices from right-to-left.
+        /// </summary>
+        public void ShowMessage(string text, float delay = 0.05f, bool alwaysScroll = false)
```
Comment (tarekgh) [inline src/devices/Max7219/MatrixTextWriter.cs]: >delay [](start = 51, length = 5)

should we use milliseconds here instead of second fraction? to be consistent with the timing APIs

File: src/devices/Max7219/MatrixTextWriter.cs (hunk 16 above)
Comment (schonbachler) [reply src/devices/Max7219/MatrixTextWriter.cs]: Depending on the application. If you want to write directly to the matrix, you can use Max7219 Device directly. If you want to write a text into the matrix, you can use the class MatrixTextWriter.

File: src/devices/Max7219/MatrixTextWriter.cs (hunk 24 above)
[the remark this answers is not in the document]
Comment (schonbachler) [reply src/devices/Max7219/MatrixTextWriter.cs]: This is a good idea for a future version

**Comment by tarekgh** [comment]
>         public static Font Default => CP437;

Just wondering, from where did you pick this font data?

---
Refers to: src/devices/Max7219/Font.cs:88 in 601f375. [](commit_id = 601f375e53cfe5cabb5eddd7aefe209cec59f79b, deletion_comment = False)

**Comment by schonbachler** [comment]
> > ```
> >     public static Font Default => CP437;
> > ```
> 
> Just wondering, from where did you pick this font data?
> 
> Refers to: src/devices/Max7219/Font.cs:88 in [601f375](https://github.com/dotnet/iot/commit/601f375e53cfe5cabb5eddd7aefe209cec59f79b). [](commit_id = [601f375](https://github.com/dotnet/iot/commit/601f375e53cfe5cabb5eddd7aefe209cec59f79b), deletion_comment = False)

adapted from [here](https://github.com/coding-world/max7219/blob/master/max7219/font.py)

**Comment by tarekgh** [comment]
> adapted from here

thanks @schonbachler, I was just curious to see the open source license on that. looks it is under MIT license. @richlander I think we are OK to pick these. right?

**Comment by tarekgh** [comment]
> @tarekgh do you have any additional comments?

I think I added some comments which not addressed yet.

File: src/devices/Max7219/Font.cs (hunk 18 above)
Comment (schonbachler) [reply src/devices/Max7219/Font.cs]: I've added a few comments and think it's clear now.

File: src/devices/Max7219/MatrixTextWriter.cs (hunk 21 above)
Comment (schonbachler) [reply src/devices/Max7219/MatrixTextWriter.cs]: Ive added some comments to the device.

**Comment by schonbachler** [comment]
> > @tarekgh do you have any additional comments?
> 
> I think I added some comments which not addressed yet.

What should I improve? Most suggestions are outdated now.

File: src/devices/Max7219/MatrixTextWriter.cs [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+using System.Linq;
+using System.Threading;
+
+namespace Iot.Device.Max7219
+{
+
+    /// <summary>
+    /// Matrix text writer can be used to write 
+    /// </summary>
+    public class MatrixTextWriter
```
Comment (tarekgh) [inline src/devices/Max7219/MatrixTextWriter.cs]: >MatrixTextWriter [](start = 17, length = 16)

I would change this name to something like MatrixGraphics (or just Graphics). the reason is, if we need to add any other graphics functionality (e.g. DrawCircle) we can just add it here. we can have the Font as a property and internally we can use default font if the user didn't specify the font in the constructor (or through the property).

Comment (schonbachler) [reply src/devices/Max7219/MatrixTextWriter.cs]: ok, renamed to MatrixGraphics

Comment (tarekgh) [reply src/devices/Max7219/MatrixTextWriter.cs]: Thanks. Do you mind adding a Font property to this class with a getter and setter?

Comment (schonbachler) [reply src/devices/Max7219/MatrixTextWriter.cs]: Already there: see line 30
```csharp
 public Font Font { get; set; }
```

Comment (tarekgh) [reply src/devices/Max7219/MatrixTextWriter.cs]: sorry I missed that. thank you.

**Comment by schonbachler** [comment]
> > ```
> >     new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 },  // 0x00
> > ```
> 
> this will be very slow initializing. did you consider just have one dimension array and do the the calculation when accessing it? if we make it one dimension array we'll allocate just one object.
> 
> sorry repeating this question but I didn't see answer or I missed the answer.
> 
> Refers to: src/devices/Max7219/Font.cs:91 in [601f375](https://github.com/dotnet/iot/commit/601f375e53cfe5cabb5eddd7aefe209cec59f79b). [](commit_id = [601f375](https://github.com/dotnet/iot/commit/601f375e53cfe5cabb5eddd7aefe209cec59f79b), deletion_comment = False)

@tarekgh something like this?
```cs
 public class FixedCharWidthFont: IFont {
    
        private readonly byte[] _data;
        private readonly int _bytesPerCharacter = 8;

        public FixedCharWidthFont(Byte[] data, int bytesPerCharacter)
        {
            _data = data;
            _bytesPerCharacter = bytesPerCharacter;
        }

        public IReadOnlyList<byte> this[char chr]
        {
            get {
                int start = chr * _bytesPerCharacter;
                int end = start + _bytesPerCharacter;
                if (end > _data.Length)
                    return Array.Empty<byte>(); //character is not defined
                else {
                    //trim
                    while (start < end && _data[start] == 0)
                        start++;
                    while (end > start && _data[end - 1] == 0)
                        end--;
                    return new ArraySegment<byte>(_data, start, end - start);
                }
            }
        }
    }
```

**Comment by tarekgh** [comment]
> something like this?

You can simplify the way of the whole font work. Instead of having a dictionary mapping from the character to the byte data of this character, you can just do, define the data as a single array like the following and define number of bytes for each characters.

```C#
        byte[] CP437 = new byte[] 
        {
            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
            0x7E, 0x81, 0x95, 0xB1, 0xB1, 0x95, 0x81, 0x7E,
            // ...
        };

       internal int BytesPerCharacter => 8;
```

Then accessing any character data would be just calculating the index ((int) ch) * BytesPerCharacter.

Let me know if this is unclear.

**Comment by tarekgh** [comment]
I forgot, you can take advantage of Span type too. something like 

```C#
internal ReadOnlySpan<byte> GetCharacterData(char c)
{
    return CP437.AsSpan().Slice((int)c * BytesPerCharacter, BytesPerCharacter );
}
```

Then the span can be used exactly as the array.

**Comment by schonbachler** [comment]
@tarekgh refactored the whole Font part.

**Comment by tarekgh** [comment]
Thanks a lot @schonbachler
